library-best-practices/glossary.md

```diff
@@ -0,0 +1,84 @@
+Java Dependency Glossary
+------------------------
+
+### Types of conflicts and compatibility
+
+- **Linkage conflict**: The signature of a non-private method or class in a
+  dependency has changed in an incompatible way between the version supplied at
+  compile time and the version invoked at runtime. For example, a public method
+  may be removed from a class or a class may be made final. Linkage conflicts
+  detected at runtime manifest as ReflectiveOperationException,
+  `NoClassDefFoundError`, `NoSuchFieldException`, `MethodNotFoundException`,
+  `LinkageError`, or other related exceptions.
+  - Or, another perspective: In cases where binary compatibility and source
+    compatibility are the same, a linkage conflict is when compilation would
+    fail if the libraries in the classpath were all built together from their
+    originating source code, or when reflection would fail.
+  - Opposite: **Linkage-compatible**.
+  - Sub-type: **Static linkage conflict**: A linkage conflict caused by a direct
+    code reference (non-reflective).
+  - Sub-type: **Reflective linkage conflict**: A linkage conflict caused by a
+    reflective reference.
+- **Behavior conflict**: The class's implementation has changed in a way that
+  can break clients at runtime although all signatures remain compatible. For
+  example, if a method that has been returning mutable lists begins returning
+  immutable lists without updating its signature, dependents that mutate the
+  list will fail, possibly far away from the original call site. By contrast, a
+  change in return type from List to ImmutableList would be a linkage conflict.
+  - Opposite: **Behavior-compatible**.
+- **Version disagreement**: This is a class of failure reported by build systems
+  with strict version checking on. The main example is dependency convergence
+  failures when using Maven version enforcer, which checks that all branches of
+  a dependency tree agree on the version of every dependency. A failure of
+  dependency convergence doesn't guarantee a linkage or behavior conflict, and
+  the passing of a dependency convergence check doesn't guarantee there will be
+  no linkage or behavior conflicts. Consequently, version disagreement is a poor
+  proxy for real problems (linkage and behavior conflicts).
+  - Dependency convergence will fail even if the dependency tree only disagrees
```

Comment:
I'm not sure the glossary should depend so heavily on the detailed rules of one plugin in one build system.

Reply:
If I remember correctly, Maven dependency convergence enforcement was originally going to form a huge part of our strategy; it was going to be the basis for our dashboard, and it was going to be part of our best practices. Now that things have changed and it is no longer part of our approach, we need to delineate it so that we can say that it is not a goal. I'm not sure how else to do that if we don't actually reference this particular plugin in this particular build system.

Reply:
Hmm that's a good general point. Should I just remove the definition of "version agreement" altogether?

Reply:
sounds like a good idea. We can always add it back in if we need it.